### PR TITLE
Add use_beachmat() helper function

### PR DESCRIPTION
**Pull request by PeteHaitch**
I initially wrote this for my personal R package, but something like it seems a useful addition to **beachmat**.

Basically automates the steps described in the 'Linking' vignette (I added a comment to that vignette).

When run from the directory of a new package (here called **myNewPackage**):

```R
beachmat::use_beachmat()
#> ✔ Adding 'Rcpp' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rhdf5lib' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'beachmat' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rcpp' to Imports field in DESCRIPTION*
#> ✔ Adding 'beachmat' to Suggests field in DESCRIPTION*
#> ✔ Adding 'HDF5Array' to Suggests field in DESCRIPTION
#> ✔ Adding 'DelayedArray' to Suggests field in DESCRIPTION
#> ✔ Adding 'Matrix' to Suggests field in DESCRIPTION
#> ✔ Setting SystemRequirements field in DESCRIPTION to 'C++11'
#> ● Include the following directives in your NAMESPACE
#> Copying code to clipboard:
#>   useDynLib('myNewPackage', .registration = TRUE)
#>   importFrom('Rcpp', 'sourceCpp')
#> ● Modify 'NAMESPACE'
#> ● Run document()
```

and finally opens `NAMESPACE` in an text editor so the user can make the required modification.

`beachmat::use_beachmat()` is modelled on `usethis::use_rcpp()` and so:

- Adds dependency on [**usethis**](http://usethis.r-lib.org/)
- Adds a `NOTE` to `R CMD check` because uses some internal functions of **usethis**
- Does sensible things if some of these steps have already been done (e.g., if package already uses **Rcpp**)
- Uses snake_case rather than camelCase (I'm not fussed either way)

BTW I haven't bumped version number.

### CODE REVIEW

**Comment by PeteHaitch** [comment]
Oh I still need to add `Makevars` and `Makevars.win`. Will add soon

File: DESCRIPTION [hunk 1]
```diff
@@ -4,12 +4,13 @@ Date: 2018-04-25
...
         "alun@wehi.edu.au"), person("Herve", "Pages", role="aut"),
-        person("Mike", "Smith", role="aut"))
+        person("Mike", "Smith", role="aut"),
+        person(c("Peter", "F."), "Hickey", role = "ctb"))
 Maintainer: Aaron Lun <alun@wehi.edu.au>
 Depends: R (>= 3.5)
 Imports: utils, methods, Rhdf5lib (>= 1.1.4), rhdf5, 
         HDF5Array (>= 1.7.3), DelayedArray (>= 0.5.30), 
-        Rcpp (>= 0.12.14) 
+        Rcpp (>= 0.12.14), usethis
```
Comment (LTLA) [inline DESCRIPTION]: Can this go into `Suggests:`?

File: R/use_beachmat.R [hunk 2]
```diff
@@ -0,0 +1,37 @@
+#' Use beachmat
+#'
+#' Creates \code{src/} and adds needed packages to \code{DESCRIPTION}.
+#'
+#' For more details see the 'Linking' vignette, available by running
+#' \code{vignette("linking", package = "beachmat")}.
+#' @importFrom usethis use_directory use_git_ignore
+#' @export
+use_beachmat <- function() {
```
Comment (LTLA) [inline R/use_beachmat.R]: Four space indenting is the norm in this package.

File: R/use_beachmat.R [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#' Use beachmat
+#'
+#' Creates \code{src/} and adds needed packages to \code{DESCRIPTION}.
+#'
+#' For more details see the 'Linking' vignette, available by running
+#' \code{vignette("linking", package = "beachmat")}.
+#' @importFrom usethis use_directory use_git_ignore
+#' @export
```
Comment (LTLA) [inline R/use_beachmat.R]: I'd prefer this not to be exported; it doesn't need to be run in dependent packages every time. If you must export it; I'm not using the Roxygen documentation tools, so you'll have to actually write a manpage manually, i.e., just delete the stuff here and remove the "Do not edit by hand" comments in the manpage.

File: man/use_beachmat.Rd [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+\title{Use beachmat}
+\usage{
+use_beachmat()
+}
+\description{
+Creates \code{src/} and adds needed packages to \code{DESCRIPTION}.
+}
+\details{
+For more details see the 'Linking' vignette, available by running
+\code{vignette("linking", package = "beachmat")}.
```
Comment (LTLA) [inline man/use_beachmat.Rd]: No manual line wrap, one line = one sentence here.

File: vignettes/linking.Rmd [hunk 5]
```diff
@@ -15,6 +15,11 @@ require(knitr)
 opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
 ```
 
+# Quick setup
```
Comment (LTLA) [inline vignettes/linking.Rmd]: Move this after or inside the Introduction. And no manual line breaks.

**Review by LTLA — CHANGES_REQUESTED** [review]

**Comment by LTLA** [comment]
Thanks Pete, added some comments for the coding style of this package.

File: DESCRIPTION (hunk 1 above)
Comment (PeteHaitch) [reply DESCRIPTION]: Done

File: R/use_beachmat.R (hunk 2 above)
Comment (PeteHaitch) [reply R/use_beachmat.R]: Fixed

File: R/use_beachmat.R (hunk 3 above)
Comment (PeteHaitch) [reply R/use_beachmat.R]: No longer exported although manpage is retained (manual).

Comment (PeteHaitch) [reply R/use_beachmat.R]: Also no longer importing anything from **usethis** into `NAMESPACE`

File: man/use_beachmat.Rd (hunk 4 above)
Comment (PeteHaitch) [reply man/use_beachmat.Rd]: Done

File: vignettes/linking.Rmd (hunk 5 above)
Comment (PeteHaitch) [reply vignettes/linking.Rmd]: Done

File: man/use_beachmat.Rd [hunk 6]
```diff
@@ -10,6 +8,5 @@ use_beachmat()
 Creates \code{src/} and adds needed packages to \code{DESCRIPTION}.
 }
 \details{
-For more details see the 'Linking' vignette, available by running
-\code{vignette("linking", package = "beachmat")}.
+For more details see the 'Linking' vignette, available by running \code{vignette("linking", package = "beachmat")}.
 }
```
Comment (LTLA) [inline man/use_beachmat.Rd]: CHECK (or BiocCheck) complains if there's no `\value` or no `\examples`. Also put yourself as `\author`.

File: vignettes/linking.Rmd [hunk 7]
```diff
@@ -28,6 +23,10 @@ Currently, the API supports double-precision, integer, logical and character mat
 Supported classes include base `matrix` objects, a number of classes from the `r CRANpkg("Matrix")` package, and disk-backed matrices from the `r Biocpkg("HDF5Array")` package.
 This document describes how to link to `r Biocpkg("beachmat")` from the C++ code of another R package.
 
+# Quick setup
+
+Run `beachmat:::use_beachmat()` in your package's directory. This will automate most of the steps described below and walk you through the remaining ones.
```
Comment (LTLA) [inline vignettes/linking.Rmd]: Newline upon sentence start.

**Review by LTLA — APPROVED** [review]

File: man/use_beachmat.Rd (hunk 6 above)
Comment (LTLA) [reply man/use_beachmat.Rd]: On hindsight, I suggest deleting this manpage entirely, as it adds absolutely nothing.

**Comment by LTLA** [comment]
Probably need a `test-use.R` where we just execute `beachmat:::use_beachmat()` in an empty dir.

**Comment by PeteHaitch** [comment]
- Added unit tests, based on tests for `usethis::use_rcpp()`. 
- Automatically generate `Makevars[.win]` if not already present

Now, when run from the directory of a new package (here called **myNewPackage**):

```R
beachmat:::use_beachmat()
#> ✔ Adding 'Rcpp' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rhdf5lib' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'beachmat' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rcpp' to Imports field in DESCRIPTION*
#> ✔ Adding 'beachmat' to Suggests field in DESCRIPTION*
#> ✔ Adding 'HDF5Array' to Suggests field in DESCRIPTION
#> ✔ Adding 'DelayedArray' to Suggests field in DESCRIPTION
#> ✔ Adding 'Matrix' to Suggests field in DESCRIPTION
#> ✔ Setting SystemRequirements field in DESCRIPTION to 'C++11'
#> ✔ Creating 'src/'
#> ✔ Adding '*.o', '*.so', '*.dll' to 'src/.gitignore'
#> ● Include the following directives in your NAMESPACE
#> Copying code to clipboard:
#>   useDynLib('myNewPackage', .registration = TRUE)
#>   importFrom('Rcpp', 'sourceCpp')
#> ● Modify 'NAMESPACE'
#> ✔ Writing 'Makevars'
#> ✔ Writing 'Makevars.win'
#> ● If your package needs to add to the $PKG_LIBS variable in Makevars and Makevars.win, please #> consult the 'Linking' vignette available via:
#>   vignette("linking", package = "beachmat")
#> ● Run document()
```

File: tests/testthat/helper.R [hunk 8]
```diff
@@ -0,0 +1,50 @@
+proj <- new.env(parent = emptyenv())
```
Comment (LTLA) [inline tests/testthat/helper.R:1]: Can't this entire file go into `test-use_beachmat.R`? It's not like any other function needs it.

Comment (PeteHaitch) [reply tests/testthat/helper.R:1]: Yeah, i think so

**Pull request merged**